applications/addons/controllers/class.addoncontroller.php

```diff
@@ -94,6 +96,109 @@ public function index($ID = '') {
     }
 
     /**
+     * Loads the current confidence record for the current addon version against
+     * the latest core version and adds it to the controller data set.
+     * 
+     * @param mixed $addon The addon from which to load the confidence record
+     * @return void
+     */
+    private function loadConfidenceRecord($addon) {
+        $session = Gdn::Session();
+        if (!$session->IsValid()) {
+            return;
+        }
+        
+        $versionID = val('AddonVersionID', $addon);
+        
+        $existingConfidenceRecord = $this->ConfidenceModel->getConfidenceVote($session->UserID, $versionID);
+        if ($existingConfidenceRecord) {
+            $this->setData('UserConfidenceRecord', $existingConfidenceRecord);
+        }
+    }
+    
+    /**
+     * Mark the current user as saying the addon works with a specific version
+     * of Vanilla
+     * 
+     * @param int $addonVersionID
+     * @param int $coreVersionID default to the latest verion of Vanilla
+     */
+    public function works($addonVersionID, $coreVersionID = false) {
+        $this->updateVote($addonVersionID, $coreVersionID, 1);
+    }
+    
+    /**
+     * Mark the current user as saying the addon is broken with a specific
+     * version of Vanilla
+     * 
+     * @param int $addonVersionID
+     * @param int $coreVersionID default to the latest verion of Vanilla
+     */
+    public function broken($addonVersionID, $coreVersionID = false) {
+        $this->updateVote($addonVersionID, $coreVersionID, 0);
+    }
+    
+    /**
+     * This is the workhorse of the confidence voting logic
+     * 
+     * Update a users vote on an addon against a specific version of Vanilla.
+     * 
+     * @param int $addonVersionID
+     * @param int $coreVersionID
+     * @param int $weight
+     * @throws NotFoundException, PermissionException
+     */
+    private function updateVote($addonVersionID, $coreVersionID, $weight) {
+        $session = Gdn::Session();
+        if (!$session->isValid()) {
+            throw permissionException('@You need to be logged in to vote.');
```

Comment:
There's a translation already for `'You must be signed in.'` which I recommend using here.